crates/ruma-common/src/events/room/encrypted/relation_serde.rs

```diff
@@ -4,37 +4,47 @@ use serde::{Deserialize, Deserializer, Serialize, Serializer};
 use super::Annotation;
 #[cfg(feature = "unstable-msc2676")]
 use super::Replacement;
+#[cfg(feature = "unstable-msc3440")]
+use super::Thread;
 use super::{InReplyTo, Reference, Relation};
+#[cfg(feature = "unstable-msc3440")]
+use crate::EventId;
 
 impl<'de> Deserialize<'de> for Relation {
     fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
     where
         D: Deserializer<'de>,
     {
-        fn convert_relation(ev: EventWithRelatesToJsonRepr) -> Relation {
-            if let Some(in_reply_to) = ev.relates_to.in_reply_to {
-                return Relation::Reply { in_reply_to };
-            }
+        let ev = EventWithRelatesToJsonRepr::deserialize(deserializer)?;
 
-            if let Some(relation) = ev.relates_to.relation {
-                return match relation {
-                    #[cfg(feature = "unstable-msc2677")]
-                    RelationJsonRepr::Annotation(a) => Relation::Annotation(a),
-                    RelationJsonRepr::Reference(r) => Relation::Reference(r),
-                    #[cfg(feature = "unstable-msc2676")]
-                    RelationJsonRepr::Replacement(Replacement { event_id }) => {
-                        Relation::Replacement(Replacement { event_id })
-                    }
-                    // FIXME: Maybe we should log this, though at this point we don't even have
-                    // access to the rel_type of the unknown relation.
-                    RelationJsonRepr::Unknown => Relation::_Custom,
-                };
+        let rel = if let Some(relation) = ev.relates_to.relation {
+            match relation {
+                #[cfg(feature = "unstable-msc2677")]
+                RelationJsonRepr::Annotation(a) => Relation::Annotation(a),
+                RelationJsonRepr::Reference(r) => Relation::Reference(r),
+                #[cfg(feature = "unstable-msc2676")]
+                RelationJsonRepr::Replacement(Replacement { event_id }) => {
+                    Relation::Replacement(Replacement { event_id })
+                }
+                #[cfg(feature = "unstable-msc3440")]
+                RelationJsonRepr::Thread(BaseThread { event_id, is_falling_back }) => {
+                    let in_reply_to = ev
+                        .relates_to
+                        .in_reply_to
+                        .ok_or_else(|| serde::de::Error::missing_field("m.in_reply_to"))?;
+                    Relation::Thread(Thread { event_id, in_reply_to, is_falling_back })
+                }
+                // FIXME: Maybe we should log this, though at this point we don't even have
+                // access to the rel_type of the unknown relation.
+                RelationJsonRepr::Unknown => Relation::_Custom,
             }
-
+        } else if let Some(in_reply_to) = ev.relates_to.in_reply_to {
+            Relation::Reply { in_reply_to }
```

Comment:
Hm, so now the relation field takes precedence over the reply field even if the relation is an unknown one. I don't think that's the right behavior.

Reply:
So we check first if it's a thread, then if it is a reply, then if it is another relation?

Reply:
Seems like a reasonabe solution.